### PR TITLE
GODRIVER-1683 Add reflection-free encoder for bson.D

### PR DESCRIPTION
**Pull request by divjotarora**

### CODE REVIEW

File: x/bsonx/noreflection_d_encoder.go [hunk 1]
```diff
@@ -0,0 +1,862 @@
...
+
+	if val.IsNil() {
+		return vw.WriteNull()
+	}
+
+	doc := val.Interface().(primitive.D)
+	return r.encodeDocument(ec, vw, doc)
+}
+
+func (r *reflectionFreeDEncoder) encodeDocumentValue(ec bsoncodec.EncodeContext, vw bsonrw.ValueWriter, v interface{}) error {
```
Comment (craiggwilson) [inline x/bsonx/noreflection_d_encoder.go:44]: I'm not sure which is odd one, but this signature has the last two parameters flipped compared to, say, the `encodeNumber` function. Can they all be the same?

File: x/bsonx/noreflection_d_encoder.go [hunk 2]
```diff
@@ -0,0 +1,862 @@
...
+	numberOk, numberErr := r.encodeNumber(ec, v, vw)
```
Comment (craiggwilson) [inline x/bsonx/noreflection_d_encoder.go:45]: Why do we have `encodeNumber` for a couple of types (but not all number types), and then this here? Why not just have them all in one switch statement?

File: x/bsonx/noreflection_d_encoder.go (hunk 1 above)
Comment (divjotarora) [reply x/bsonx/noreflection_d_encoder.go:44]: Thanks for catching this. I changed them all to be `(ec, vw, valueToEncode)`.

File: x/bsonx/noreflection_d_encoder.go (hunk 2 above)
Comment (divjotarora) [reply x/bsonx/noreflection_d_encoder.go]: I don't remember why I thought `encodeNumber` was useful. I removed the function and moved all of its cases to the type-switch below.

**Pull request merged**